// @neo4j/graphql/dist/classes/Neo4jGraphQL.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const schema_1 = require("@graphql-tools/schema");
const graphql_parse_resolve_info_1 = require("graphql-parse-resolve-info");
const schema_2 = require("../schema");
const utils_1 = require("../utils");
const index_1 = require("../auth/index");
class Neo4jGraphQL {
    constructor(input) {
        this.driver = input.driver;
        this.driverConfig = input.driverConfig;
        const { nodes, schema } = schema_2.makeAugmentedSchema({
            typeDefs: input.typeDefs,
            resolvers: input.resolvers,
            schemaDirectives: input.schemaDirectives,
        });
        if (input.debug) {
            // eslint-disable-next-line no-console
            let logger = console.log;
            if (typeof input.debug === "function") {
                logger = input.debug;
            }
            this.debug = (message) => logger(message);
        }
        this.nodes = nodes;
        this.schema = this.createWrappedSchema({ schema, driver: input.driver, driverConfig: input.driverConfig });
        this.document = graphql_1.parse(graphql_1.printSchema(schema));
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,class-methods-use-this
    debug(message) {
        return undefined;
    }
    createWrappedSchema({ schema, driver, driverConfig, }) {
        return schema_1.addSchemaLevelResolver(schema, (_obj, _args, context, resolveInfo) => {
            /*
                Deleting this property ensures that we call this function more than once,
                See https://github.com/ardatan/graphql-tools/issues/353#issuecomment-499569711
            */
            // eslint-disable-next-line no-param-reassign,no-underscore-dangle
            delete resolveInfo.operation.__runAtMostOnce;
            if (!(context === null || context === void 0 ? void 0 : context.driver)) {
                if (!driver) {
                    throw new Error("A Neo4j driver instance must either be passed to Neo4jGraphQL on construction, or passed as context.driver in each request.");
                }
                context.driver = driver;
            }
            if (!(context === null || context === void 0 ? void 0 : context.driverConfig)) {
                context.driverConfig = driverConfig;
            }
            context.neoSchema = this;
            context.resolveTree = graphql_parse_resolve_info_1.parseResolveInfo(resolveInfo);
            context.jwt = index_1.getJWT(context);
        });
    }
    checkNeo4jCompat(input = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const driver = input.driver || this.driver;
            const driverConfig = input.driverConfig || this.driverConfig;
            if (!driver) {
                throw new Error("neo4j-driver Driver missing");
            }
            return utils_1.checkNeo4jCompat({ driver, driverConfig });
        });
    }
}
exports.default = Neo4jGraphQL;
